### PR TITLE
Alpha Backend Framework 🚀 

### PR DESCRIPTION
**Pull request by jhaals**
Initial implementation of the Backstage backend framework

Further details into the structure and reasoning can be found in #11611

### CODE REVIEW

File: plugins/catalog-node/package.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "@backstage/catalog-node",
+  "description": "The catalog-node module for @backstage/plugin-catalog-backend",
+  "version": "0.0.0",
+  "main": "src/index.ts",
+  "types": "src/index.ts",
+  "license": "Apache-2.0",
+  "private": true,
```
Comment (Rugvip) [inline plugins/catalog-node/package.json:8]: Nope 😁

File: packages/backend-app-api/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# @backstage/backend-app-api
+
+This package provides the core API used by Backstage backend apps.
```
Comment (Rugvip) [inline packages/backend-app-api/README.md:3]: Let's add a fat disclaimer at the top of these that it's all experimental and should not be used

File: packages/backend/package.json [hunk 3]
```diff
@@ -26,6 +26,8 @@
     "build-image": "docker build ../.. -f Dockerfile --tag example-backend"
   },
   "dependencies": {
+    "@backstage/backend-plugin-api": "^0.0.0",
+    "@backstage/backend-app-api": "^0.0.0",
```
Comment (Rugvip) [inline packages/backend/package.json:30]: I'm thinking we wire this up in a new `backend-next` package instead?

File: packages/backend/src/next/index.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+//     });
+//   },
+// });
+
+const backend = createBackend({
+  apis: [],
+});
+
+backend.add(catalogPlugin({}));
+backend.add(scaffolderCatalogExtension({}));
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/backend/src/next/index.ts]: TBD :grin:. Most likely stuff like that will happen though services (Utility API equivalent), as they will have access to add root routes

File: plugins/catalog-backend/package.json [hunk 5]
```diff
@@ -34,6 +34,9 @@
     "clean": "backstage-cli package clean"
   },
   "dependencies": {
+    "@backstage/backend-app-api": "^0.0.0",
```
Comment (Rugvip) [inline plugins/catalog-backend/package.json:37]: The catalog plugin should not need access to the app API, only the plugin one

File: packages/backend-plugin-api/src/wiring/types.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+export interface BackendPluginConfig<TOptions> {
+  id: string;
+  register(reg: BackendInitRegistry, options: TOptions): void;
+}
+
+export function createBackendPlugin<TOptions>(
+  config: BackendPluginConfig<TOptions>,
+): (option: TOptions) => BackendRegistrable {
```
[the remark this answers is not in the document]
Comment (jhaals) [reply packages/backend-plugin-api/src/wiring/types.ts:81]: Yes we should definitely do that 👍

File: packages/backend-next/package.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "example-backend-next",
+  "version": "0.2.73-next.2",
```
Comment (Rugvip) [inline packages/backend-next/package.json:3]: Fix

File: packages/backend-plugin-api/package.json [hunk 8]
```diff
@@ -0,0 +1,53 @@
+{
+  "name": "@backstage/backend-plugin-api",
+  "description": "Core API used by Backstage backend plugins",
+  "version": "0.0.0",
```
Comment (Rugvip) [inline packages/backend-plugin-api/package.json:4]: Add changesets for new packages

File: packages/backend-plugin-api/src/services/definitions/scheldulerServiceRef.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+import { createServiceRef } from '../system/types';
+import {
+  PermissionAuthorizer,
+  PermissionEvaluator,
+} from '@backstage/plugin-permission-common';
+
+/**
+ * @public
+ */
+export const permissionsServiceRef = createServiceRef<
```
Comment (Rugvip) [inline packages/backend-plugin-api/src/services/definitions/scheldulerServiceRef.ts:26]: I feel 🍻 was involved here

File: packages/backend-plugin-api/src/wiring/types.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    options: TOptions,
+  ): void;
+}
+
+export function createBackendModule<TOptions>(
+  config: BackendModuleConfig<TOptions>,
+): (option: TOptions) => BackendRegistrable {
+  return options => ({
+    id: `${config.pluginId}.${config.moduleId}`,
+    register(register) {
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/backend-plugin-api/src/wiring/types.ts:106]: Yep need a separate type here for modules most likely, future fix though

**Review by Rugvip — APPROVED** [review]

**Pull request merged**